File: my-folder/0054-spiral-matrix/solution.java
class Solution {
    public List<Integer> spiralOrder(int[][] matrix) {
        List<Integer> res = new ArrayList<>();
        int rows = matrix.length;
        if(rows==0){
            return res;
        }
        int cols = matrix[0].length;
        int r1 = 0;
        int r2 = rows-1;
        int c1 = 0;
        int c2 = cols-1;
        while(r1<=r2&&c1<=c2){
            // right
            for(int i=c1;i<=c2;++i){
                res.add(matrix[r1][i]);
            }
            r1++;
            
            // down
            for(int i=r1;i<=r2;++i){
                res.add(matrix[i][c2]);
            }
            c2--;
            
            if(r1>r2||c1>c2){
                break;
            }
            // left
            for(int i=c2;i>=c1;--i){
                res.add(matrix[r2][i]);
            }
            r2--;
            
            // up
            for(int i=r2;i>=r1;--i){
                res.add(matrix[i][c1]);
            }
            c1++;
        }
        
        return res;
    }
}

File: my-folder/0054-spiral-matrix/solution.py
class Solution:
    def spiralOrder(self, matrix: List[List[int]]) -> List[int]:
        m,n = len(matrix), len(matrix[0])
        visited = []
        visited_node = 0
        cur = [0,0]
        dirs = [[0,1],[1,0],[0,-1],[-1,0]]
        dir_num = 0
        ans = []
        while visited_node < m*n:
            visited_node += 1
            i,j = cur[0],cur[1]
            visited.append((i,j))
            ans.append(matrix[i][j])
            cur[0], cur[1] = i + dirs[dir_num][0], j + dirs[dir_num][1]
            if cur[0] < 0 or cur[0] >= m or cur[1] < 0 or cur[1] >= n or (cur[0],cur[1]) in visited:
                dir_num = (dir_num + 1) % 4
                cur[0], cur[1] = i + dirs[dir_num][0], j + dirs[dir_num][1]
        return ans
            
        
            
        
